Bump mono to -std=gnu11

lambdageek opened the pull request:
Change the mono build to require C11 (with gnu extensions on gcc/clang platforms).
* Change `g_static_assert` to be `_Static_assert` or `static_assert` as apropriate.
* Change `_DN_STATIC_ASSERT` to be `static_assert`
* Add static asserts in `jiterp.c` when it casts between `T*` and `atomic_T*`
* Add C11 guidance to the mono coding guide doc

Contributes to #90404

lambdageek committed:
Bump mono to -std=gnu99; use static_assert

lambdageek committed:
Add an obviously bad static assertion

lambdageek committed:
don't fall back to runtime checks for g_static_assert

lambdageek committed:
Revert "Add an obviously bad static assertion"

This reverts commit c588bd9c1fdb4ab9010a8ad98db696176d001ba5.

lambdageek committed:
fix static assert that wasn't a constant expression

lambdageek committed:
use static_assert in shared containers

lambdageek committed:
bump C standard in offsets-tool.py

jkoritzinsky commented on src/mono/CMakeLists.txt:
```diff
@@ -525,7 +525,7 @@ if(GCC)
   else()
     # We require C99 with some GNU extensions, e.g. `linux` macro
-    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99")
+    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu11")
```
Instead of passing the standard flag directly, can we change this to set the CMake properties that emit these versions (`CMAKE_C_STANDARD_REQUIRED`, `CMAKE_C_STANDARD`, and `CMAKE_C_EXTENSIONS`)?

lambdageek replied:
yes, CMAKE_C_EXTENSIONS is what I didn't know about, thanks!

And that means we can actually just inherit `CMAKE_C_STANDARD_REQUIRED` and `CMAKE_C_STANDARD` from `eng/native/configurecompiler.cmake`.  Nice.

am11 replied:
> ah nevermind we don't include that in Mono yet.

Yup, once it's done, it will cleanup some workarounds such as: https://github.com/dotnet/runtime/blob/c540d01de6a2656722cbe85b9fa3eac400510279/src/native/libs/System.IO.Compression.Native/extra_libs.cmake#L3 and cleanup duplicate code from cmake files under src/mono.

lambdageek committed:
use _Static_assert before C23

Dont' include assert.h in glib.h because some of our older 3P code
includes assert.h on its own and there are conflicts

lambdageek committed:
use CMAKE_C_STANDARD and related properties

lambdageek committed:
jiterp: static_assert that atomic ops are (less likely) to go wrong

Not every C implementation guarantees that atomic operations on
arbtrary types are lock free.  So for example, casting between atomic_ushort* and
uint16_t* might not actually be ok to do.  We can't assert that
they're inter-castable, but at least assert that they're the same size
and that atomic_ushort is always lock-free. There might still be
restrictions (for example atomic_ushort might have to be aligned
differently) but this should at least catch obvious data corruption.

lambdageek commented on src/mono/mono/mini/interp/jiterpreter.c:
```diff
@@ -925,6 +928,7 @@ mono_interp_tier_prepare_jiterpreter_fast (
 #ifdef DISABLE_THREADS
 	gint64 count = trace_info->hit_count++;
 #else
+	static_assert (sizeof(atomic_long) == sizeof(gint64) && ATOMIC_LONG_LOCK_FREE == 2);
 	gint64 count = atomic_fetch_add ((atomic_long *)&trace_info->hit_count, 1);
```
@kg PTAL at this file.

This particular use of `atomic _long` is actually kind of suspect.  on some platforms it's actually `long long` (and `atomic_llong`) that is 64-bits.  I assume that's not the case on wasm?

kg replied:
Yeah this might need to be atomic_llong. I did test it originally but it's possible it's truncating it.

lambdageek replied:
```
/__w/1/s/src/mono/mono/mini/interp/jiterpreter.c:931:2: error: static assertion failed due to requirement 'sizeof(_Atomic(long)) == sizeof (trace_info->hit_count)': 
          static_assert (sizeof(atomic_long) == sizeof(trace_info->hit_count) && ATOMIC_LONG_LOCK_FREE == 2, "");
          ^              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /usr/local/emscripten/emsdk/upstream/emscripten/cache/sysroot/include/assert.h:12:23: note: expanded from macro 'static_assert'
  #define static_assert _Static_assert
                        ^
  /__w/1/s/src/mono/mono/mini/interp/jiterpreter.c:931:37: note: expression evaluates to '4 == 8'
```

yep. looks like it needs to be `atomic_llong`

lambdageek committed:
Add C11 guidance to the Mono coding guide

lambdageek commented on docs/coding-guidelines/mono-code-guide.md:
```diff
@@ -61,6 +61,38 @@ There are actually three boolean types to keep in mind:
 * `MonoBoolean` used as an interop type with C# bool in internal calls
 * `mono_bool` used by the public C API - generally new code shouldn't use it except when adding a new public API function.
 
+## Newer features of C11 and later
```
I added some coding guidelines based on what I know right now. Feedback welcome

lambdageek committed:
clarify that these rules don't apply to pinvokes outside the runtime

AaronRobinsonMSFT commented on src/mono/mono/mini/interp/jiterpreter.c:
```diff
@@ -751,11 +753,12 @@ trace_info_get (gint32 index) {
 #else
-	gint32 index = atomic_fetch_add ((atomic_int *)&trace_count, 1),
+	static_assert (sizeof(atomic_int) == sizeof(gint32) && ATOMIC_INT_LOCK_FREE == 2);
+	gint32 index = atomic_fetch_add ((atomic_int *)&trace_count, 1);
```
Would it be more desirable to use the target instead of the target's type for the assert?

For example:

```c
static_assert (sizeof(atomic_int) == sizeof(trace_count) && ATOMIC_INT_LOCK_FREE == 2);
```

lambdageek committed:
jiterp: fix static_assert; use sizeof(expr) instead of sizeof(type)

lambdageek committed:
jiterp: long is 32-bits on wasm; use llong

AaronRobinsonMSFT reviewed (approved):
LGTM
